### PR TITLE
Gradle 7.1

### LINKED ISSUE #975: Gradle building
Why Gradle is better then Maven? It has easy plugin interface and flexible build scripts. You can write custom tasks and include scripts...

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65905608-gradle-building?utm_campaign=plugin&utm_content=tracker%2F14691067&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F14691067&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by VaiTon**
Closes https://github.com/GlowstoneMC/Glowstone/issues/975

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+tasks.jar {
+    enabled = false
+    manifest {
+        attributes(
+                "Launcher-Agent-Class" to "net.glowstone.util.ClassPathAgent",
+                "Main-Class" to "net.glowstone.GlowServer",
+
+                "Implementation-Title" to project.name,
+                "Implementation-Version" to "${project.version}.git-${project.name}-${getGitHash()}",
```
Comment (smartboyathome) [inline build.gradle.kts:103]: This seems fine for pipeline versions, but when developing, the version built could override the commit you're building upon? Basically, the issue Git doesn't assign a hash for uncommitted changes.

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    id("com.github.johnrengelman.shadow") version "7.0.0"
+}
+
+val buildExtras = System.getenv("BUILD_EXTRAS")?.toBoolean() ?: false
+val apiVersion = "1.16.5-R0.1-SNAPSHOT"
+
+repositories {
+    mavenLocal()
+    mavenCentral()
```
Comment (smartboyathome) [inline build.gradle.kts]: We don't pull dependencies from Maven Central, our repo.glowstone.net repository instead proxies to it, sot his should probably be removed.

**Review by smartboyathome — CHANGES_REQUESTED** [review]
Those two changes are my only gripes with this, but I would prefer if @gdude2002 also reviewed this.

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    maven("https://repo.glowstone.net/repository/maven-public/")
+    maven("https://repo.glowstone.net/repository/snapshots/")
+    maven("https://libraries.minecraft.net")
+}
+
+dependencies {
```
Comment (gdude2002) [inline build.gradle.kts:22]: This won't do - use [version catalogs](https://docs.gradle.org/current/userguide/platforms.html#sub:version-catalog).

We make use of them at Kord Extensions, and it's much, much easier to manage deps this way. For example:

* [settings.gradle.kts](https://github.com/Kord-Extensions/kord-extensions/blob/root/settings.gradle.kts)
* [libs.version.toml](https://github.com/Kord-Extensions/kord-extensions/blob/root/libs.versions.toml)
* [build.gradle.kts](https://github.com/Kord-Extensions/kord-extensions/blob/root/kord-extensions/build.gradle.kts)

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.1-bin.zip
```
Comment (gdude2002) [inline gradle/wrapper/gradle-wrapper.properties]: Nitpick, but may as well use 7.1.1.

**Review by gdude2002 — CHANGES_REQUESTED** [review]
Pretty good, just a couple of things. You can consider this a "+1" to the other review comments as well.

File: build.gradle.kts (hunk 3 above)
Comment (VaiTon) [reply build.gradle.kts:22]: I have a question regarding this. How is the support of GitHub dependabot on this? I would prefer putting versions in settings.gradle.kts and let dependabot update them. What do you think?

File: build.gradle.kts (hunk 1 above)
Comment (VaiTon) [reply build.gradle.kts:103]: I don't seem to understand how could this be a problem :/

File: build.gradle.kts (hunk 3 above)
Comment (gdude2002) [reply build.gradle.kts:22]: Gradle support in dependabot is already fairly lacking, although I do wonder if it's that much of a deal in our case - I don't find myself getting many emails for it here, if any

Comment (VaiTon) [reply build.gradle.kts:22]: I'm doing it! :P

File: build.gradle.kts (hunk 1 above)
Comment (smartboyathome) [reply build.gradle.kts:103]: It just could give someone the wrong idea about what version a given implementation was based on if it was built locally.

**Review by gdude2002 — APPROVED** [review]

**Review by smartboyathome — APPROVED** [review]

**Pull request merged**